build-aux/build-index.js

```diff
@@ -0,0 +1,310 @@
+#!/usr/bin/env -S gjs -m
+
+import Gio from "gi://Gio";
+import GLib from "gi://GLib";
+
+Gio._promisify(
+  Gio.File.prototype,
+  "load_contents_async",
+  "load_contents_finish",
+);
+
+Gio._promisify(
+  Gio.FileEnumerator.prototype,
+  "next_files_async",
+  "next_files_finish",
+);
+
+Gio._promisify(
+  Gio.File.prototype,
+  "enumerate_children_async",
+  "enumerate_children_finish",
+);
+
+Gio._promisify(
+  Gio.File.prototype,
+  "replace_contents_async",
+  "replace_contents_finish",
+);
+
+// Biblioteca is GTK4 only; these are GTK 3 libraries
+// we only support gi-docgen and gtk3 uses gtk-doc
+const IGNORED_LIBRARIES = ["atk", "libhandy-1"];
+
+const SECTION_TYPES = {
+  class: ["Classes", "#classes"],
+  content: ["Addition Documentation", "#extra"],
+  interface: ["Interfaces", "#interfaces"],
+  record: ["Structs", "#structs"],
+  alias: ["Aliases", "#aliases"],
+  enum: ["Enumerations", "#enums"],
+  bitfield: ["Bitfields", "#bitfields"],
+  function: ["Functions", "#functions"],
+  function_macro: ["Function Macros", "#function_macros"],
+  domain: ["Error Domains", "#domains"],
+  callback: ["Callbacks", "#callbacks"],
+  constant: ["Constants", "#constants"],
+};
+
+const SUBSECTION_TYPES = {
+  ctor: ["Constructors", "#constructors"],
+  type_func: ["Functions", "#type-functions"],
+  method: ["Instance Methods", "#methods"],
+  property: ["Properties", "#properties"],
+  signal: ["Signals", "#signals"],
+  class_method: ["Class Methods", "#class-methods"],
+  vfunc: ["Virtual Methods", "#virtual-methods"],
+};
+
+const REQUIRED = ["class", "interface", "record", "domain"];
+const INDEX = [];
+
+await loadDocs();
+
+async function loadDocs() {
+  await Promise.all([
+    scanLibraries(Gio.File.new_for_path("/app/share/doc")),
+    scanLibraries(Gio.File.new_for_path("/app/share/doc/glib-2.0")),
+  ]);
+  sort_index(INDEX);
+
+  const [pkgdatadir] = ARGV;
+  GLib.mkdir_with_parents(pkgdatadir, 0o755);
+
+  await Gio.File.new_for_path(pkgdatadir)
+    .get_child("doc-index.json")
+    .replace_contents_async(
+      new TextEncoder().encode(JSON.stringify(INDEX)),
+      null,
+      false,
+      Gio.FileCreateFlags.NONE,
+      null,
+    );
+}
+
+async function scanLibraries(base_dir) {
+  const libraries = [];
+
+  const iter = await base_dir.enumerate_children_async(
+    "standard::name,standard::type",
+    Gio.FileQueryInfoFlags.NOFOLLOW_SYMLINKS,
+    GLib.PRIORITY_DEFAULT,
+    null,
+  );
+  // eslint-disable-next-line no-constant-condition
+  while (true) {
+    const infos = await iter.next_files_async(10, GLib.PRIORITY_DEFAULT, null);
+    if (infos.length === 0) break;
+
+    for (const info of infos) {
+      if (info.get_file_type() !== Gio.FileType.DIRECTORY) continue;
+
+      if (IGNORED_LIBRARIES.includes(info.get_name())) continue;
+
+      const directory = iter.get_child(info);
+      libraries.push(loadLibrary(directory).catch(console.error));
+    }
+  }
```

Comment:
I thought the idea behind this was so that we can read 10 files parallelly at a time which is much faster doing 1 file at a time? I understand that it won't matter much since this code is being executed during build time but we should try and speed up the build process as well, no?

Reply:
Are you sure this is always the case? I tried testing it myself and I'm getting inconsistent timings. At times the current method would perform better but other times your changes give better results. Just to confirm, could you run this a couple of times to see if the results are consistent? If it doesn't matter much, we'll use your changes because it's much simpler.